### PR TITLE
Breaks the moveBy call from drag and dragFrom into two separate calls and changes the default behavior of DragStartBehavior to DragStartBehavior.start

### PR DESCRIPTION
**Pull request by jslavitz**
For easier fixing of tests related to #26246, drag moveBy calls are broken in two and changes the default DragStartBehavior in all widgets with drag recognizers to DragStartBehavior.start.

This PR fixes the issue caused by introducing the DragStartBehavior.start in #26246 where the dragUpdate callback from a drag recognizer would never be called. By breaking the moveby call into two parts, the first move will trigger the drag recognizer to accept the drag gesture and the second drag will invoke the dragUpdated callback with the total delta for this moveBy. Each component of the new moveBy calls is in the same direction as the original call and the sum magnitude of all the moveBy calls is exactly equal to the magnitude of the original move by call.

### CODE REVIEW

File: packages/flutter_test/lib/src/controller.dart [hunk 1]
```diff
@@ -424,11 +433,29 @@ abstract class WidgetController {
...
+        await gesture.moveBy(Offset(offset.dx - (xSign * (touchSlopX - eps)), 0));
+      } else {
+        await gesture.moveBy(Offset(offset.dx, 0));
+      }
+      if (offset.dy * ySign > touchSlopY + eps) {
+        await gesture.moveBy(Offset(0, ySign * (touchSlopY + eps)));
+        await gesture.moveBy(Offset(0, offset.dy - (ySign * (touchSlopY - eps))));
+      } else {
+        await gesture.moveBy(Offset(0, offset.dy));
+      }
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter_test/lib/src/controller.dart]: yup, I already fixed this

File: packages/flutter_test/lib/src/controller.dart [hunk 2]
```diff
@@ -414,8 +414,17 @@ abstract class WidgetController {
   ///
   /// If you want the drag to end with a speed so that the gesture recognition
   /// system identifies the gesture as a fling, consider using [fling] instead.
-  Future<void> drag(Finder finder, Offset offset, { int pointer }) {
-    return dragFrom(getCenter(finder), offset, pointer: pointer);
+  ///
+  /// {@template flutter.flutter_test.lib.src.controller}
+  /// By default, if the x or y component of offset is greater than kTouchSlop, the
+  /// gesture is broken up into two separate moves calls. Changing [touchSlopX] or
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter_test/lib/src/controller.dart]: Looks like you missed one...

File: packages/flutter_test/lib/src/controller.dart [hunk 3]
```diff
@@ -420,11 +429,30 @@ abstract class WidgetController {
   /// If you want the drag to end with a speed so that the gesture recognition
   /// system identifies the gesture as a fling, consider using [flingFrom]
   /// instead.
-  Future<void> dragFrom(Offset startLocation, Offset offset, {int pointer}) {
+  ///
+  /// {@macro flutter.flutter_test.lib.src.controller}
```
Comment (Hixie) [inline packages/flutter_test/lib/src/controller.dart]: looks like this is pointing at the wrong template now

File: packages/flutter_test/lib/src/controller.dart [hunk 4]
```diff
@@ -420,11 +429,30 @@ abstract class WidgetController {
...
+  Future<void> dragFrom(Offset startLocation, Offset offset, { int pointer, double touchSlopX = kTouchSlop, double touchSlopY = kTouchSlop }) {
     return TestAsyncUtils.guard<void>(() async {
       final TestGesture gesture = await startGesture(startLocation, pointer: pointer);
       assert(gesture != null);
-      await gesture.moveBy(offset);
+
+      final double xSign = offset.dx.sign;
+      final double ySign = offset.dy.sign;
+
+      final bool separateX = offset.dx * xSign > touchSlopX;
```
Comment (Hixie) [inline packages/flutter_test/lib/src/controller.dart]: generally we use `math.abs` to get the absolute value

File: packages/flutter_test/lib/src/controller.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter_test/lib/src/controller.dart]: Is there a test that verifies that this is fixed?

In particular, it should verify at least the following cases (assuming kTouchSlop is 10 to make the maths simpler):

- dragging 3,8 - one drag
- dragging 8,3 - one drag.
- dragging 20,5 - 2 drags, one by 10,2.5, next by another 10,2.5.
- dragging 20,15 - 3 drags, one by 10,7.5, next by 3.333,2.5, finally another by the remainder (6.666,5).
- dragging 15,20 - 3 drags, one by 7.5,10, next by 2.5,3.333, one by 5,6.666.
- dragging 20,20 - 2 drags, one by 10,10, next by another 10,10.
- dragging 5,20 - 2 drags, one by 2.5,10, next by another 2.5,10.
- dragging 0,5 - 1 drag
- dragging -150,200 - 3 drags, one by -75,100, next by -25,33.333, one by -50,66.666.

[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter_test/lib/src/controller.dart]: Literally square intersection 😮

[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter_test/lib/src/controller.dart]: That last example is wrong.

File: packages/flutter/test/cupertino/refresh_test.dart [hunk 5]
```diff
@@ -124,15 +124,15 @@ void main() {
       verify(mockHelper.builder(
         any,
         RefreshIndicatorMode.drag,
-        50.0,
+        32.08099300425676,
```
Comment (Hixie) [inline packages/flutter/test/cupertino/refresh_test.dart]: For future reference (no need to change anything in this patch): It's generally a bad sign when you have a non-round number like this in a test. Normally, the numbers in a test should have obvious sources: e.g. they should be computed on the fly, or they should be round numbers that can obviously be derived from other round numbers, etc.

The problem with a number like this one is that as a reviewer I have no idea if it's correct. Is it 32.08099300425676? Or should it be 32.08099310425676? Or 32.08099300425672? There's no real way for me to know. This means that when the test is changed again, I've no way to know if the test was originally wrong, or if the person updating the test made a mistake.

Comment (Hixie) [reply packages/flutter/test/cupertino/refresh_test.dart]: see also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-mysterious-and-magical-numbers-that-lack-a-clear-derivation which I just added

File: packages/flutter/test/cupertino/refresh_test.dart [hunk 6]
```diff
@@ -325,15 +325,15 @@ void main() {
...
+            132.08099300425673,
             100.0, // Default value.
             60.0, // Default value.
           ),
           mockHelper.refreshTask(),
           mockHelper.builder(
             any,
             RefreshIndicatorMode.armed,
-            argThat(moreOrLessEquals(127.10396988577114)),
+            argThat(moreOrLessEquals(111.92012371530531)),
```
Comment (Hixie) [inline packages/flutter/test/cupertino/refresh_test.dart]: this is a great example. I have no way to review this to know if it's correct or not.

File: packages/flutter/test/widgets/list_view_horizontal_test.dart [hunk 7]
```diff
@@ -35,7 +35,7 @@ void main() {
     await tester.pumpWidget(buildFrame(textDirection: TextDirection.ltr));
 
     await tester.pump(const Duration(seconds: 1));
-    await tester.drag(find.text('1'), const Offset(-300.0, 0.0));
+    await tester.drag(find.text('1'), const Offset(-318.0, 0.0));
```
Comment (Hixie) [inline packages/flutter/test/widgets/list_view_horizontal_test.dart]: This seems wrong. Nothing should have changed here: there's nothing competing with the drag gesture.

File: packages/flutter_test/test/controller_test.dart [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+void main() {
+  testWidgets(
+    'WidgetTester.drag must break the offset into multiple parallel components if'
+    'the drag goes outside the touch slop values',
+    (WidgetTester tester) async {
+      // The first Offset in every sub array (ie. offsetResults[i][0]) is (touchSlopX, touchSlopY).
+      // The second Offset ... (ie. offsetResults[i][1]) will be the total move offset.
+      // The remaining values ... are the expected separated drag offsets.
+      final List<List<Offset>> offsetResults = <List<Offset>>[
```
Comment (Hixie) [inline packages/flutter_test/test/controller_test.dart]: trivial nit: for consistency, the literals that look like int literals should have trailing `.0`s.

File: packages/flutter_test/lib/src/controller.dart [hunk 9]
```diff
@@ -420,11 +429,76 @@ abstract class WidgetController {
...
+              await gesture.moveBy(Offset(offsetX - signedSlopX, offsetY - diffY2 - signedSlopY));
+            }
+          }
+        } else { // The drag goes through the corner of the box.
+          await gesture.moveBy(Offset(signedSlopX, signedSlopY));
+          await gesture.moveBy(Offset(offsetX - signedSlopX, offsetY - signedSlopY));
+        }
+      } else { // The drag ends inside the box.
+        await gesture.moveBy(offset);
+      }
```
Comment (Hixie) [inline packages/flutter_test/lib/src/controller.dart]: This code is very clean and well commented.

**Comment by Hixie** [comment]
I'm a little concerned about some of the tests that were affected that don't look like they should have been. Other than that, this looks pretty good.

File: packages/flutter/lib/src/widgets/scroll_activity.dart [hunk 10]
```diff
@@ -364,12 +363,14 @@ class ScrollDragController implements Drag {
...
+
     offset = _adjustForScrollStartThreshold(offset, details.sourceTimeStamp);
     if (offset == 0.0) {
       return;
     }
     if (_reversed) // e.g. an AxisDirection.up scrollable
       offset = -offset;
+
     delegate.applyUserOffset(offset);
   }
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_activity.dart]: what's with the changes in this file?

File: packages/flutter/test/cupertino/date_picker_test.dart [hunk 11]
```diff
@@ -487,7 +487,8 @@ void main() {
         ),
       );
 
-      await tester.drag(find.text('March'), const Offset(0.0, 32.0));
+      await tester.drag(find.text('March'), const Offset(0, 32.0), touchSlopY: 0);
+
```
Comment (Hixie) [inline packages/flutter/test/cupertino/date_picker_test.dart]: please use `0.0` for the double zero literal

Comment (Hixie) [reply packages/flutter/test/cupertino/date_picker_test.dart]: (also, the new blank line seems unnecessary)

File: packages/flutter/test/cupertino/date_picker_test.dart [hunk 12]
```diff
@@ -525,15 +526,14 @@ void main() {
...
+      await tester.drag(find.text('27'), const Offset(0.0, -32.0), touchSlopY: 0);
       await tester.pump();
       expect(
         date,
         DateTime(2018, 2, 28),
       );
 
-
-      await tester.drag(find.text('28'), const Offset(0.0, -32.0));
+      await tester.drag(find.text('28'), const Offset(0.0, -32.0), touchSlopY: 0);
```
Comment (Hixie) [inline packages/flutter/test/cupertino/date_picker_test.dart]: same comment about 0.0 here and other places

File: packages/flutter/test/cupertino/date_picker_test.dart [hunk 13]
```diff
@@ -721,14 +721,14 @@ void main() {
       expect(date, DateTime(2018, 1, 1, 11, 59));
 
       // 11:59 -> 9:59
-      await tester.drag(find.text('11'), -_kRowOffset * 2);
+      await tester.drag(find.text('11'), -_kRowOffset * 1.5);
```
Comment (Hixie) [inline packages/flutter/test/cupertino/date_picker_test.dart]: why do we change both the multiplier and the value? this seems to suggest the math is wrong. I don't understand why we change both.

File: packages/flutter/test/cupertino/date_picker_test.dart [hunk 14]
```diff
@@ -721,14 +721,14 @@ void main() {
...
       await tester.pump();
       await tester.pump(const Duration(milliseconds: 500));
 
       expect(date, DateTime(2018, 1, 1, 9, 59));
 
       // 9:59 -> 15:59
-      await tester.drag(find.text('9'), _kRowOffset * 6);
+      await tester.drag(find.text('9'), _kRowOffset * 4.2);
```
Comment (Hixie) [inline packages/flutter/test/cupertino/date_picker_test.dart]: ditto

File: packages/flutter/test/cupertino/refresh_test.dart [hunk 15]
```diff
@@ -482,6 +482,7 @@ void main() {
         100.0, // Default value.
         60.0, // Default value.
       ));
+
```
Comment (Hixie) [inline packages/flutter/test/cupertino/refresh_test.dart]: why the new blank line?

File: packages/flutter/test/cupertino/switch_test.dart [hunk 16]
```diff
@@ -270,21 +269,21 @@ void main() {
 
     expect(value, isFalse);
 
-    await tester.drag(find.byType(CupertinoSwitch), const Offset(-30.0, 0.0));
+    await tester.drag(find.byType(CupertinoSwitch), const Offset(-48.0, 0.0), touchSlopX: 19);
```
Comment (Hixie) [inline packages/flutter/test/cupertino/switch_test.dart]: why change the offset _and_  the touch slop?

File: packages/flutter_test/test/controller_test.dart [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+import 'package:flutter/gestures.dart';
+
+void main() {
+  testWidgets(
+    'WidgetTester.drag must break the offset into multiple parallel components if'
+    'the drag goes outside the touch slop values',
+    (WidgetTester tester) async {
+      // The first Offset in every sub array (ie. offsetResults[i][0]) is (touchSlopX, touchSlopY).
+      // The second Offset ... (ie. offsetResults[i][1]) will be the total move offset.
+      // The remaining values ... are the expected separated drag offsets.
```
Comment (Hixie) [inline packages/flutter_test/test/controller_test.dart]: what's with the "..."s?

File: packages/flutter_test/test/controller_test.dart [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+        expect(
+          offsetResults[resultI].length - 2,
+          dragOffsets.length,
+
+          reason:
+            'There is a difference in the number of expected and actual split offsets for the drag with:\n'
+            'Touch Slop: ' + testResult[0].toString() + '\n'
+            'Delta:      ' + testResult[1].toString() + '\n',
+        );
+        for (int valueI = 2; valueI < offsetResults[resultI].length; valueI += 1) {
```
Comment (Hixie) [inline packages/flutter_test/test/controller_test.dart]: what's the "I" in "resultI" and "valueI"?

Comment (jslavitz) [reply packages/flutter_test/test/controller_test.dart]: Index

**Comment by Hixie** [comment]
Notes from our discussion:
- the new drag should drag _past_ kTouchSlop
- since that's 18, it makes sense to drag to 20, that way the maths will be simpler
- make sure to assert that the default is greater than kTouchSlop
- we should update the documentation for kTouchSlop to be unambiguous on this point
- we should update the documentation for the drag test method to explain what's going on

File: packages/flutter/lib/src/gestures/constants.dart [hunk 19]
```diff
@@ -54,6 +54,11 @@ const Duration kZoomControlsTimeout = Duration(milliseconds: 3000);
 /// the gesture is a scroll gesture, or, inversely, the maximum distance that a
 /// touch can travel before the framework becomes confident that it is not a
 /// tap.
+///
+/// Mathematically speaking, if a touch moves with a total delta in the range
+/// [0, 'kTouchSlop'] it is not considered to be a drag, whereas if the total
+/// delta is in the range ('kTouchSlop', Infinity), it is considered to be a
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/constants.dart]: this is probably clearer if written as "... a total delta less than or equal to [kTouchSlop] it is not considered to be a drag, whereas if the total delta is greater than [kTouchSlop], it is considered..."

That also avoids any risk that dartdoc will get confused by the `[...]`

**Comment by Hixie** [comment]
cc @goderbauer for final review

File: packages/flutter_test/lib/src/controller.dart [hunk 20]
```diff
@@ -410,8 +414,18 @@ abstract class WidgetController {
   ///
   /// If you want the drag to end with a speed so that the gesture recognition
   /// system identifies the gesture as a fling, consider using [fling] instead.
-  Future<void> drag(Finder finder, Offset offset, {int pointer}) {
-    return dragFrom(getCenter(finder), offset, pointer: pointer);
+  ///
+  /// {@template flutter.flutter_test.drag}
+  /// By default, if the x or y component of offset is greater than [kTouchSlop], the
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/controller.dart]: When reading this doc I'm left with a few unanswered questions. Maybe you can add some more information to answer them for our curious users?

* Why are the gestures broken up into multiple gestures in the first place?
* Why should I ever modify touchSlopX or tochSlopY?
* If I change them how do I determine the values I want for them?

**Comment by goderbauer** [comment]
nit: It would probably be helpful for future archeologists (and current reviewers) to add a little more context to the PR description: Why are we breaking moveBy calls into two? How do we break them up? Why does this make fixing tests easier? etc.

What's also missing from the description (and probably the title of this PR) is that this is actually changing the default behavior from DragStartBehavior.down to DragStartBehavior.start. This is a breaking change, no? Can you add the "breaking change" label? I believe communications for this breaking change has already been send out a while ago? Might be worth reminding people on that thread that this change is happening now if that was send out a long time ago.

File: packages/flutter_test/lib/src/controller.dart [hunk 21]
```diff
@@ -420,11 +445,76 @@ abstract class WidgetController {
...
+        final double inverseOffsetSlope = offsetX / offsetY;
+        final double slopSlope = touchSlopY / touchSlopX;
+        final double absoluteOffsetSlope = offsetSlope.abs();
+        final double signedSlopX = touchSlopX * xSign;
+        final double signedSlopY = touchSlopY * ySign;
+        if (absoluteOffsetSlope != slopSlope) {
+          // The drag goes through one or both of the extents of the edges of the box.
+          if (absoluteOffsetSlope < slopSlope) {
+            // The drag goes through the vertical edge of the box.
+            // It is guaranteed that the |offsetX| > touchSlopX.
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/controller.dart]: Can this be expressed as an assert instead?

**Review by goderbauer — APPROVED** [review]
LGTM

Thanks for adding the context!

File: packages/flutter_test/lib/src/controller.dart (hunk 21 above)
Comment (jslavitz) [reply packages/flutter_test/lib/src/controller.dart]: yup!

**Pull request merged**